### ISSUE TITLE
Entity Reference Revisions are Growing Out of Control

### ISSUE THREAD

**Issue by bryanpizzillo** [NCIOCPL/cgov-digital-platform#2815]
### Issue description
The ERR module version we use does not clean up itself. https://www.drupal.org/project/entity_reference_revisions/issues/2771531. Upgrading to 1.7 fixes part of the issue. An additional issue is that PDQ recreates the sections on load, so combined with this bug our database is growing about a gig every few months.

So we need to figure out how to keep the entity revisions in check. 

> **ESTIMATE** NN

**Comment by bryanpizzillo** [comment]
Pasting in a slack chat from 6/23 so all is not lost on the questions and possible things to account for.

**Slack Chat on the Subject**
mingao  2 hours ago
paragraph_revision__field_pdq_section_html




bryanpizzillo  2 hours ago
How large is that?




bryanpizzillo  2 hours ago
Like all of the extra space, or just bigger than most (edited) 




mingao  2 hours ago
paragraph__field_pdq_section_html and paragraph_revision__field_pdq_section_html similar in size, each is more than 2GB




mingao  1 hour ago
let me check some backup of last year to compare




mingao  1 hour ago
backups for Oct. shows the above two tables were of size 1GB of each




mingao  1 hour ago
number of rows this month is about 97340 for paragraph_revision__field_pdq_section_html, while in Oct. it was 36854




bryanpizzillo  1 hour ago
Yeah so revisions are being stored biggly (edited) 




bryanpizzillo  1 hour ago
what about the other revision tables?




bryanpizzillo  1 hour ago
anything super large?




mingao  1 hour ago
not exactly, since the revision table is almost same size of paragraph__field_pdq_section_html




mingao  1 hour ago
basically for table: paragraph__field_pdq_section_html




mingao  1 hour ago
3425 out of 96624 rows are used by node__field_summary_sections. While 93199 rows does not seem to be used by node__field_summary_sections




mingao  1 hour ago
I think I mentioned that same problem to you before @bryanpizzillo




mingao  1 hour ago
basically, paragraph tables not deleted when node__field* are deleted




mingao  1 hour ago
I encountered this problem for reports of impact analysis




bryanpizzillo  1 hour ago
I remember with impact analysis… I did not know that it would be to the order of a gig ever few months!




bryanpizzillo  1 hour ago
that is a REALLY bad bug




bryanpizzillo  1 hour ago
Ok, so what I am hearing is the following:
PDQ Revisions are not removed and therefore grow
PDQ Sections do not actually get deleted when they are updated, when a section is updated a new row in the paragraph__field pdq_section_html table is added - leaving behind the old one and a revision?




mingao  1 hour ago
yes




bryanpizzillo  1 hour ago
@bkline are you catching any of this?




mingao  1 hour ago
tables like paragraphs_item_field_data also not deleted and also any other paragraph tables, but only summary table is this visible, same problem for article section too.




mingao  1 hour ago
remove a paragraph from its parent only removes from node__field* tables, none of paragraph tables get deleted.




bryanpizzillo  44 minutes ago
is there any changes on that issue?




mingao  41 minutes ago
https://www.drupal.org/project/paragraphs/issues/2935630   this one does not say any change




bryanpizzillo  40 minutes ago
It looks like if we upgrade ERR to 1.7 there is a manual cleanup tool




mingao  39 minutes ago
https://www.drupal.org/project/paragraphs/issues/2741937?




bryanpizzillo  38 minutes ago
https://www.drupal.org/project/entity_reference_revisions/releases/8.x-1.7




bryanpizzillo  38 minutes ago
This release includes several improvements on cleaning up orphaned composite entities (e.g. paragraphs) when deleting host entities, translations and revisions. Additionally, there is a manual cleanup process that will scan all referenced entities and delete revisions and entities that are no longer used.




bryanpizzillo  37 minutes ago
There is also a fix for https://www.drupal.org/node/2953650




bryanpizzillo  36 minutes ago
Actually, I also found https://www.drupal.org/project/paragraphs/issues/2741937




bryanpizzillo  35 minutes ago
TBH these could be OTHER issues…




bryanpizzillo  34 minutes ago
Here is the meta issue for all the orphaning problems : https://www.drupal.org/project/entity_reference_revisions/issues/2771531




bkline  24 minutes ago
@bkline are you catching any of this?
every word




bkline  22 minutes ago
I even figured out ERR




bryanpizzillo  21 minutes ago
To ERR is Paragraphs…




bkline  21 minutes ago
how true




bryanpizzillo  21 minutes ago
we need shirts made I think
:smirk:
1





mingao  20 minutes ago
so based on https://www.drupal.org/project/entity_reference_revisions/issues/2771531, it says "Not used in earlier revisions of the host entity". In our situation, those paragraphs are used by earlier revisions




bkline  20 minutes ago
that sounds right




bkline  19 minutes ago
I wondered if the paragraphs really were orphaned




mingao  19 minutes ago
e.g. a summary has 4035 revision




bryanpizzillo  19 minutes ago
Well, I figure this should fix part of our issues, right? We may also need to put PDQ on a diet




mingao  18 minutes ago
Childhood Cancer Genomics (PDQ®)–Health Professional Version (en and es combined number)




bkline  18 minutes ago
I've always wondered if we really needed to revision the PDQ content




bryanpizzillo  18 minutes ago
(or any revisions — like keep only 6mo of revisions or 2 previous…)




bkline  17 minutes ago
I'm game




bkline  17 minutes ago
but for PDQ what's the use case for keeping ANY superseded revisions? (edited) 




bryanpizzillo  16 minutes ago
Dunno really — the only thing I can think of is that something goes horribly wrong with a CDR publish and we need to revert the summaries that are blank while you all look for the root cause




bryanpizzillo  15 minutes ago
this has actually happened in prod more than once




bryanpizzillo  15 minutes ago
Everything was sent to GK as normal




bkline  15 minutes ago
but not in the Drupal world, right?




bryanpizzillo  15 minutes ago
the problem was that some summaries with a bunch of empty sectiosn got send




bryanpizzillo  14 minutes ago
No not yet, but this was actually a filter issue IIRC




bryanpizzillo  14 minutes ago
Luckily in percussion we were able to revert




bryanpizzillo  14 minutes ago
NOW




bkline  14 minutes ago
OK, so maybe what we do with ES




bkline  13 minutes ago
keep the rolling last N revisions




bryanpizzillo  13 minutes ago
if we apply this fix, the prevision version will be completely removed, right?




bkline  13 minutes ago
???




bryanpizzillo  13 minutes ago
Because once the item is removed from field_pdq_section_html isn’t that saying that it should be deleted?




bryanpizzillo  12 minutes ago
IDK how exactly they fixed things




bkline  11 minutes ago
which "fix" are we talking about exactly -- you and Min were discussing a bunch of things that have been fixed




mingao  11 minutes ago
I see, Bryan is saying table paragraph__field_pdq_section_html  should be trimed greatly after the fix




mingao  10 minutes ago
while paragraph_revision__field_pdq_section_html can still have some previous revision?




bryanpizzillo  7 minutes ago
@bkline When we update to ERR 1.7, from that moment on “Composite entities” should be cleaned up. Now I don’t know if that means when the “host” node or container paragraph is deleted, or what…




bryanpizzillo  5 minutes ago
In my mind there is field_sections  or some such that points to an entity for the paragraph. That table also has a set of revisions of all the previous paragraphs that have existed on that summary.  Then there are the field_pdq_section_html for the HTML content of the paragraphs pointed to by field_section. (edited) 




bryanpizzillo  4 minutes ago
There is a paragraph entity table too…




bkline  4 minutes ago
well, I guess we can apply the update to a lower tier,  see if the tool does a good enough job without clobbering something we need, and go with it if satisfied




mingao  4 minutes ago
paragraphs_item_field_data?




bryanpizzillo  2 minutes ago
So we should not delete anything from the field_pdq_section_html until the corresponding row is deleted from paragraph entity. Paragraph entity row should not be deleted until it is removed from field_sections_revisions … or else you can’t roll back…

**Comment by bkline** [comment]
Added [wiki page](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Drupal-Storage-for-PDQ-Summaries) describing how Drupal stores the PDQ summaries.

**Comment by bryanpizzillo** [comment]
Entity Reference Revisions module has been updated as part of #2905.

**Comment by bkline** [comment]
## Proposed approach for pruning older PDQ CIS revisions/orphaned summary sections
* Create a REST API end point (`PATCH /pdq/api`) to remove older revisions of specific nodes. The body of the request contains a JSON object with a `nodes` member consisting of an array of node ID integers, and an optional `keep` member with an integer indicating the maximum number of revisions to keep for each language present for the node (by default the most recent three published revisions are preserved for each language). The response for such a request takes the form `[[node-id, [revision-id, ...]], ...]` representing the revisions which have been dropped.
* Create a REST API end point (`PATCH /pdq/api/cis`) to remove orphaned `paragraph` entities (that is, entities whose parents no longer exist, or which never were associated with a parent) as well as orphaned revisions for `paragraph` entities which also have revisions whose parents still exist. The body of the request contains a single required integer specifying the maximum number of deletions to be performed. The implementation first attempts to identify that number of orphaned `paragraph` entities and delete them. If that step results in fewer deletions than were requested, the software then looks for orphaned revisions in the remaining `paragraph` entities, deleting as many of them as necessary to effect as many deletions as requested (if possible), but no more. The response to such a request is a two-integer array `[number-of-entities-deleted, number-of-separate-revisions-deleted]`.
* Create a `drush` command `cgov:prune-node-revisions` which can be used to drop older revisions of specified nodes (or of all of the nodes of a specified node type), similar to the functionality described for the first REST API end point above. As with that API, the number of per-language published revisions to preserve can be overridden with a command-line option.
* Create a `drush` command `cgov:drop-orphaned-summary-sections` which can be used to delete orphaned `paragraph` entities and revisions, similar to the second REST API end point described above. The maximum number of deletions to be performed can be specified on the command line, defaulting to all orphans.
* Use the `drush` command to clear out the accumulated backlog of unwanted older revisions and orphaned summary sections.
* Add invocations of the two new REST API end points to the CDR publishing software, the first invoked in batches of node IDs until all of the nodes affected by the publishing job have been pruned, and the second invoked in batches until the Drupal server reports that no more orphaned summary sections remain to be deleted.

## Proposed testing for the new functionality
1. Copy the production database tables to a non-production server.
2. Run a report identifying which revisions of which PDQ CIS nodes should remain after a complete pruning run.
3. Capture the JSON for each published PDQ Cancer Information Summary using `GET /pdq/api/cis/{id}`.
4. Use the `drush cgov:prune-node-revisions` command to drop all unwanted older PDQ CIS revisions.
5. Run a report identifying which summary section `paragraph` entities with which revisions should remain after a complete pruning.
6. Run the `drush cgov:drop-orphaned-summary-sections` to delete all orphaned summary sections.
7. Verify that the set of revisions which exist for PDQ CIS nodes matches the report from step 2.
8. Verify that the set of summary section `paragraph` entities/revisions matches the report from step 5.
9. Capture the JSON for each published PDQ Cancer Information Summary using `GET /pdq/api/cis/{id}` and verify that it matches the JSON captured in step 4.
10. Republish all of the PDQ Cancer Information Summaries.
11. Verify that the latest two revisions from the report in step 2 (assuming there were at least two revisions) plus a new revision are present for each of the summaries.
12. Verify that no orphaned summary sections remain.

**Comment by bkline** [comment]
I may tinker with the idea of adding a 13th step to the testing, pulling down the JSON for each PDQ CIS a third time and comparing with the earlier two, if I can find an easy way to mask out expected deltas (posted dates changing, _etc._).

**Comment by bryanpizzillo** [comment]
> Create a drush command cgov:drop-orphaned-summary-sections which can be used to delete orphaned paragraph entities and revisions, using the second new service described above. The maximum number of deletions to be performed can be specified on the command line, defaulting to all orphans

Can this be (easily) named pdq:drop-orphaned-summary-sections since it is PDQ only?

**Comment by bryanpizzillo** [comment]
> Create a new service in the pdq_core module to identify older revisions of a specific node which can be deleted, preserving only the most recent N published revisions for each language present for the node, where N defaults to 3.

Tell me more, this affects a comment about `cgov:prune-node-revisions`. Will you delete any paragraphs entities that were referenced in any ERR fields on the node that are no longer referenced by any remaining revision?

If the answer is no, then can `cgov:prune-node-revisions` become `pdq:prune-summary-revisions` and only work for the pdq_cancer_information_summary node type? If it does not clean up the paragraphs, then it is a dangerous command for any node type since it would leave orphans. It sounds like the process for a manual cleanup of PDQ summaries is to run:
```
drush cgov:prune-node-revisions <summary ids>
drush pdq:drop-orphaned-summary-sections
```

Correct? If so, there are no parallel orphan cleanup for the other fields and types.

**Comment by bryanpizzillo** [comment]
> Use the drush command to clear out the accumulated backlog of unwanted older revisions and orphaned summary sections.

Command or commands? I assume you run both commands.

**Comment by bkline** [comment]
I have made all the changes above that you requested, and I will modify the `pdq:prune-summary-revisions` command to restrict its use to PDQ nodes. And you are correct: at the moment we are not implementing paragraph orphan cleanup for the non-PDQ content.

**Comment by bkline** [comment]
I have run all of the checks described above twice on my local workstation, as well as on ODE-401, and everything appears to be working as it should. @zhuomingao is also doing some independent analysis of the before/after database tables to see if there are any problems. I have attached the [tools and detailed instructions](https://github.com/NCIOCPL/cgov-digital-platform/files/5823448/testing-for-issue-2815.zip) for what I did to verify the correctness of the software.
